### PR TITLE
egressgateway: provide a very basic Cell

### PR DESCRIPTION
**Pull request by lmb**
These are the minimal changes needed to get the EGW into a hive Cell. The hacky part is that enabling the EGW is gated on the global option.Config.EnableIPv4EgressGateway which is references from a bunch of places. It's not entirely clear to me how to migrate that toggle yet, and it'll probably require a bunch of changes. I've kept the PR simple in the interests of making this easier to review.

### CODE REVIEW

**Comment by lmb** [comment]
/test-1.16-4.19
/test-1.26-net-next
/test-runtime

**Comment by lmb** [comment]
cc @jibi @joamaki

File: pkg/egressgateway/manager.go [hunk 1]
```diff
@@ -45,6 +54,20 @@ const (
 	eventDeleteEndpoint
 )
 
+type Config struct {
+	// Install egress gateway IP rules and routes in order to properly steer
+	// egress gateway traffic to the correct ENI interface
+	InstallRoutes bool
+}
+
+var DefaultConfig = Config{
```
Comment (bimmlerd) [inline pkg/egressgateway/manager.go]: Does this have to be exported? Would prefer it to be priv so it's not possible to mutate it from outside

File: pkg/egressgateway/manager.go [hunk 2]
```diff
@@ -32,6 +34,13 @@ var (
 	zeroIPv4 = net.ParseIP("0.0.0.0")
 )
 
+var Cell = cell.Module(
```
Comment (bimmlerd) [inline pkg/egressgateway/manager.go:39]: Since this is what code navigation will jump you to as an entry point for this file, would be nice to have a comment here explaining what this cell provides

File: pkg/egressgateway/manager.go [hunk 3]
```diff
@@ -32,6 +34,13 @@ var (
...
+	"egressgateway",
+	"IPv4 Egress Gateway",
```
Comment (bimmlerd) [inline pkg/egressgateway/manager.go]: this is usually a one line description of what the cell does

File: pkg/egressgateway/manager.go [hunk 4]
```diff
@@ -81,18 +104,51 @@ type Manager struct {
...
+
+type Params struct {
+	cell.In
+
+	Config            Config
+	CacheStatus       k8s.CacheStatus
+	IdentityAllocator identityCache.IdentityAllocator
+
+	// Lifecycle to attach hooks to. Optional.
+	Lifecycle hive.Lifecycle
```
Comment (bimmlerd) [inline pkg/egressgateway/manager.go:116]: Is this optional for testing's sake? Seems cleaner to me to construct a hive and get a lifecycle from it in the test.

File: daemon/cmd/cells.go [hunk 5]
```diff
@@ -91,6 +92,18 @@ var (
 
 		// IPCache, policy.Repository and CachingIdentityAllocator.
 		cell.Provide(newPolicyTrifecta),
+
+		// Conditionally instantiate the Egress Gateway.
+		cell.Config(egressgateway.DefaultConfig),
+		cell.Provide(
+			func(params egressgateway.Params) *egressgateway.Manager {
+				if !option.Config.EnableIPv4EgressGateway {
```
Comment (bimmlerd) [inline daemon/cmd/cells.go]: if there's not too much churn maybe pull this into the egress config and out of option.config. But this might be what you alluded to re minimal changes; fine in that case

**Review by bimmlerd — COMMENTED** [review]
couple of questions and some docs nit

File: daemon/cmd/cells.go (hunk 5 above)
Comment (lmb) [reply daemon/cmd/cells.go]: yes, this is what i was referring to. I had a chat with gilberto and maybe there is an easy way to work around this, i'll take another look.

File: pkg/egressgateway/manager.go (hunk 4 above)
Comment (lmb) [reply pkg/egressgateway/manager.go:116]: It's optional so that I don't have to change all testing call sites, is that what you mean? Integrating hive into the tests causes the PR to balloon again, for little direct benefit IMO.

File: pkg/egressgateway/manager.go (hunk 1 above)
Comment (lmb) [reply pkg/egressgateway/manager.go]: It's exported because of the ugly "enable egw" hack, I'll try to get rid of it.

File: pkg/egressgateway/manager.go (hunk 4 above)
Comment (bimmlerd) [reply pkg/egressgateway/manager.go:116]: that's fine, in that case maybe mention that it's optional for testing, and should be supplied in the usual case

File: daemon/cmd/cells.go (hunk 5 above)
Comment (lmb) [reply daemon/cmd/cells.go]: I've replaced a bunch of checks for `Config.Enable..` with `egressGatewayManager != nil`. That works nicely for the k8swatcher, the daemon is less clear cut. I've put the latter in a separate commit.

We're still left with a bunch of hairy global usage unfortunately:
```
pkg/datapath/loader/base.go
343:	if option.Config.Tunnel == option.TunnelDisabled && option.Config.EnableIPv4EgressGateway {

pkg/k8s/synced/crd.go
56:	if option.Config.EnableIPv4EgressGateway {

pkg/k8s/watchers/cilium_endpoint.go
212:	if option.Config.EnableIPv4EgressGateway {
239:	if option.Config.EnableIPv4EgressGateway {

pkg/datapath/linux/config/config.go
270:	if option.Config.EnableIPv4EgressGateway {
```

I think to fix this in a nicer way we need a way to define flags in hive cells, but somehow plumb them into `option.Config` during the transition phase. Otherwise we need to do gigantic refactors again. cc @joamaki

File: pkg/egressgateway/manager.go (hunk 4 above)
Comment (lmb) [reply pkg/egressgateway/manager.go:116]: > Integrating hive into the tests causes the PR to balloon again, for little direct benefit IMO.

Turns out i was full of crap, it's fine. Reworked to start a hive.

File: daemon/cmd/cells.go (hunk 5 above)
Comment (joamaki) [reply daemon/cmd/cells.go]: Preferably I'd of course see these in `EgressGatewayConfig`, but seeing where the enable flag is referenced it'd get painful to do that in this PR since you'd need to plumb the EgressGatewayConfig struct to all these places that have not yet been modularized. Let's keep that out from this PR for the time being. 

As a transitionary thing we could already add EgressGatewayConfig, keep the option.Config.EnableIPv4EgressGateway, remove the flag registration from `daemon/cmd/daemon_main.go` and do it in EgressGatewayConfig.Flags() instead, and finally set `option.Config.EnableIPv4EgressGateway` somewhere... and the last point is where it gets a bit hairy. It can be made to work by having a `cell.Invoke` at the very top in `daemon/cmd/cells.go` which would depend on `*DaemonConfig` and `EgressGatewayConfig` and bridge the two before the "daemonCell" or anyone else gets access to it. I don't like it that much though. I think it'd be preferable to just do it cleanly once when we're far enough that we can do it.

Another thing here is that most likely `EgressGatewayConfig` is a "datapath" configuration, e.g. it affects the operation of datapath (above e.g. pkg/datapath/loader/base.go) more than control-plane and it should likely live under `pkg/datapath` and possibly stay internal to it. Control-plane components that manage the "egress gateway datapath" would then depend on the config either directly, or via some "egress gateway datapath API" to tell them what can and cannot be done. This is maybe less clear with egress gateway, but with datapath components that require probing to figure out the "actual" configuration, it makes sense to not expose the user provided config struct, but rather the configuration that's the result of merging user configuration and probe results. 

tl;dr: let's do nothing in this PR about the config :grin:

**Review by joamaki — APPROVED** [review]

File: daemon/cmd/daemon.go [hunk 6]
```diff
@@ -683,8 +685,9 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup,
 
 	d.cgroupManager = manager.NewCgroupManager()
 
-	if option.Config.EnableIPv4EgressGateway {
-		d.egressGatewayManager = egressgateway.NewEgressGatewayManager(cacheStatus, d.identityAllocator, option.Config.InstallEgressGatewayRoutes)
+	var egressGatewayWatcher watchers.EgressGatewayManager
+	if d.egressGatewayManager != nil {
+		egressGatewayWatcher = d.egressGatewayManager
```
Comment (jibi) [inline daemon/cmd/daemon.go:694]: Can't we just pass `d.egressGatewayManager` to `watchers.NewK8sWatcher()`?

**Review by jibi — APPROVED** [review]

File: daemon/cmd/daemon.go (hunk 6 above)
Comment (lmb) [reply daemon/cmd/daemon.go:694]: Unfortunately not, because of the dreaded `nil` interface bug. Passing directly will construct the following:

```go
var v watchers.EgressGatewayManager = (*Manager)(nil)
// we need
var v watchers.EgressGatewayManager = nil
```

Want me to comment that?

Comment (jibi) [reply daemon/cmd/daemon.go:694]: Makes sense, no need to add a comment

File: pkg/egressgateway/manager.go (hunk 4 above)
Comment (bimmlerd) [reply pkg/egressgateway/manager.go:116]: I might be missing something, but I see neither a hive nor a non-optional Lifecycle - did you back out of this again?

Comment (lmb) [reply pkg/egressgateway/manager.go:116]: Yes, because hive was blowing up on me in CI a couple of times and I couldn't figure out what was wrong. Jussi has since helped me, but overall I'm not convinced hive is a net gain for tests. It moves a bunch of logic to runtime that in tests is much easier to catch at compile time.

Comment (bimmlerd) [reply pkg/egressgateway/manager.go:116]: Alright, fair enough.

**Review by bimmlerd — APPROVED** [review]

**Comment by lmb** [comment]
/test-1.26-net-next

**Review by youngnick — APPROVED** [review]

**Comment by lmb** [comment]
This is what CI driven development looks like! The test-1.24-5.4 is, I believe, unrelated. Jenkins doesn't even have logs for the cilium pods and one of @jibi PRs has the same failure.

**Comment by pchaigno** [comment]
> The test-1.24-5.4 is, I believe, unrelated. Jenkins doesn't even have logs for the cilium pods and one of @jibi PRs has the same failure.

Do we have an issue for that?

**Comment by jibi** [comment]
Filled https://github.com/cilium/cilium/issues/24667 for the CI failure
